sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+
+**Rationale:** This method is used to facilitate interactions with off-chain
+services, and to export data from a smart contract that can be consumed
+off-chain.
+
+**Examples:**
+
+```clarity
+(to-consensus-buff 1) ;; Returns (some 0x0000000000000000000000000000000001)
+(to-consensus-buff u1) ;; Returns (some 0x0100000000000000000000000000000001)
+(to-consensus-buff true) ;; Returns (some 0x03)
+(to-consensus-buff false) ;; Returns (some 0x04)
+(to-consensus-buff none) ;; Returns (some 0x09)
+(to-consensus-buff 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR) ;; Returns (some 0x051fa46ff88886c2ef9762d970b4d2c63678835bd39d)
+(to-consensus-buff { abc: 3, def: 4 }) ;; Returns (some 0x0c00000002036162630000000000000000000000000000000003036465660000000000000000000000000000000004)
+```
+
+### New method: `from-consensus-buff`
+
+* **Input Signature:** `(from-consensus-buff type-signature buff)`
+* **Output Signature:** `(optional t)`
+
+The `from-consensus-buff` function is a special function that will deserialize a
+buffer into a Clarity value, using the SIP-005 serialization of the
+Clarity value. The type that `from-consensus-buff` tries to deserialize
+into is provided by `type-signature`. If it fails
+to deserialize the `buff` argument to this type, the method returns `none`.
+
+**Rationale:** This method is used to facilitate interactions with off-chain
+services, and to implement "transaction-less transactions.""  The ability to
+both verify the signature on a buffer and decode the buffer into a Clarity
+value is a basic building block for bridges and oracles.  In addition, it is a
+basic building block for smart contracts where users are not expected to send
+transactions (e.g. because they lack a wallet, or because the transaction volume
+would be too high).  Instead, they can send signed Clarity values to someone willing to
+package them into transactions at a later time (possibly as a batch), and
+send them to the smart contract on their behalf.
+
+**Examples:**
+
+```clarity
+(from-consensus-buff int 0x0000000000000000000000000000000001) ;; Returns (some 1)
+(from-consensus-buff uint 0x0000000000000000000000000000000001) ;; Returns none
+(from-consensus-buff uint 0x0100000000000000000000000000000001) ;; Returns (some u1)
+(from-consensus-buff bool 0x0000000000000000000000000000000001) ;; Returns none
+(from-consensus-buff bool 0x03) ;; Returns (some true)
+(from-consensus-buff bool 0x04) ;; Returns (some false)
+(from-consensus-buff principal 0x051fa46ff88886c2ef9762d970b4d2c63678835bd39d) ;; Returns (some SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR)
+(from-consensus-buff { abc: int, def: int } 0x0c00000002036162630000000000000000000000000000000003036465660000000000000000000000000000000004) ;; Returns (some (tuple (abc 3) (def 4)))
+```
+
+### New method: `replace-at`
+
+* **Input Signature:** `(replace-at (x (sequence Y)) (i uint) (value Y))`
+* **Output Signature:** `(optional (sequence Y))`
```

Comment:
What will be the outcome of the following:

```clarity
(replace-at "abcdefg", u2 "xyz")
```
`abxyzdefg` or `abxdefg` or `abxyzfg` ?

Despite the fact that this question seems trivial, all 3 answers can be valid. I've seen different implementations of such function so it is better to specify its outcome now and this example with "proper" answer.

If it is the first one, then description must be modified, because return type is not the same as the input.

Reply:
None of the above -- it wouldn't type-check (and it has a syntax error).  The third argument is a 1-item sequence (for `buff`, `string-ascii`, or `string-utf8`) or has the type of an element of the sequence (for `list`), so you'd need `(replace-at "abcdefg" u2 "x")` or something.

Reply:
Understand. Thank you!

English is not my first language, but I think it could be describe a little bit more explicitly ie. 
> The `replace-at` function takes in a sequence, an index, and a replacement value, and returns a new sequence with the data at the index position replaced with the replacement value. The given replacement value's type must match the type of a single element of input sequence. The return type on success is the same type as the input sequence (both type and length wise). The sequence can be a list, buffer, string-ascii, or string-utf8. The replacement value can be a list element, 1 byte long buffer, single character string-ascii, or single character string-utf8.